lib/gc/ExecutionEngine/CPURuntime/Memorypool.cpp

```diff
@@ -0,0 +1,213 @@
+    b = next;
+  }
+}
+
+size_t filoMemoryPool::getBlockSize(size_t sz) const {
+  // calculate the aligned size of management blocks in the header
+  constexpr size_t header_size =
+      divideAndCeil(sizeof(memoryBlock) + sizeof(memoryChunk),
+                    defaultAlignment) *
+      defaultAlignment;
+  // the allocated size should include the aligned header size
+  sz = sz + header_size;
+  if (sz > blockSize_) {
+    return divideAndCeil(sz, getOsPageSize()) * getOsPageSize();
+  } else {
+    return blockSize_;
+  }
+}
+
+void *filoMemoryPool::alloc(size_t sz) {
+  if (unlikely(!buffers_)) {
+    buffers_ = memoryBlock::make(getBlockSize(sz), nullptr, nullptr);
+    current_ = buffers_;
+  }
+  do {
+    intptr_t newptr = current_->callAllocPtr();
+    size_t newallocated = newptr + sz - reinterpret_cast<intptr_t>(current_);
+    if (likely(newallocated <= current_->size_)) {
+      // if the current block is not full
+      size_t alloc_size = newallocated - current_->allocated_;
+      current_->allocated_ = newallocated;
+      memoryChunk *chunk = memoryChunk::init(newptr, alloc_size);
+      return reinterpret_cast<void *>(newptr);
+    }
+    // if the block is full, check the next block
+    // if there is no next block left, allocate a new one
+    if (!current_->next_) {
+      current_->next_ = memoryBlock::make(getBlockSize(sz), current_, nullptr);
+    }
+    current_ = current_->next_;
+  } while (true);
+}
+
+void filoMemoryPool::dealloc(void *ptr) {
+  auto intptr = reinterpret_cast<intptr_t>(ptr);
+  auto intcur = reinterpret_cast<intptr_t>(current_);
+  // Optional: check if the pointer is valid in the current block
+
+  assert(intptr > intcur &&
+         intptr - intcur < static_cast<ptrdiff_t>(current_->size_));
+  auto chunk = reinterpret_cast<memoryChunk *>(intptr - sizeof(memoryChunk));
+  // Optional: check if the stack is ok
+  assert(chunk->canary_ == memoryChunk::magicCheckNum_ &&
+         "Corrupt stack detected");
+  assert(current_->allocated_ > chunk->size_);
+  current_->allocated_ -= chunk->size_;
+
+  // skip the empty blocks
+  while (unlikely(current_->allocated_ == sizeof(memoryBlock))) {
+    if (current_->prev_) {
+      current_ = current_->prev_;
+    } else {
+      break;
+    }
+  }
+}
+
+void filoMemoryPool::release() {
+  freeMemoryBlockList(buffers_);
+  buffers_ = nullptr;
+  current_ = nullptr;
+}
+
+void filoMemoryPool::clear() {
+  for (auto cur = current_; cur; cur = cur->prev_) {
+    cur->allocated_ = sizeof(memoryBlock);
+  }
+  current_ = buffers_;
+}
+
+filoMemoryPool::~filoMemoryPool() { release(); }
+
+static thread_local filoMemoryPool mainMemoryPool_{memory_pool::mainChunkSize};
+// if the current thread is a worker thread, use this pool
+static thread_local filoMemoryPool threadMemoryPool_{threadlocalChunkSize};
+
+} // namespace memory_pool
+} // namespace gc
+} // namespace mlir
+
+extern "C" void *gcAlignedMalloc(size_t sz) noexcept {
+  if (sz == 0) {
```

Comment:
why do we need this? If it is really needed, you can wrap `sz==0` with `unlikely(...)` for better performance.

Reply:
This aligns with the legacy implementation. Maybe we can delete this.